pkg/target/manager.go

```diff
@@ -0,0 +1,527 @@
+		// update the target path to the shoot; this is primarily important
+		// when so far neither project nor seed were set. By updating the
+		// target, we persist the result of the resolving step earlier and make
+		// it easier to other gardenctl commands to ingest the target without
+		// having to re-resolve the shoot name again.
+		// resolveShootName will only ever return either a project or a seed,
+		// never both. The decision what to prefer happens there as well.
+		if project != nil {
+			t.Project = project.Name
+		}
+
+		if seed != nil {
+			t.Seed = seed.Name
+		}
+
+		if err := m.updateShootKubeconfig(ctx, gardenClient, t, shoot); err != nil {
+			return fmt.Errorf("failed to fetch kubeconfig for shoot cluster: %w", err)
+		}
+
+		return nil
+	})
+}
+
+// resolveShootName takes a shoot name and tries to find the matching shoot
+// on the given garden. Either project or seed can be supplied to help in
+// finding the Shoot. If no or multiple Shoots match the given criteria, an
+// error is returned.
+// If a project or a seed are given, they are returned directly (unless an
+// error is returned). If neither are given, the function will decide how
+// to best find the Shoot later by returning either a project or a seed,
+// never both.
+func (m *managerImpl) resolveShootName(
+	ctx context.Context,
+	gardenClient client.Client,
+	project *gardencorev1beta1.Project,
+	seed *gardencorev1beta1.Seed,
+	shootName string,
+) (*gardencorev1beta1.Project, *gardencorev1beta1.Seed, *gardencorev1beta1.Shoot, error) {
+	shoot := &gardencorev1beta1.Shoot{}
+
+	// If a shoot is targeted via a project, we fetch it based on the project's namespace.
+	// If the target uses a seed, _all_ shoots in the garden are filtered
+	// for shoots with matching seed and name.
+	// If neither project nor seed are given, _all_ shoots in the garden are filtered by
+	// their name.
+	// It's an error if no or multiple matching shoots are found.
+
+	if project != nil {
+		// fetch shoot from project namespace
+		key := types.NamespacedName{Name: shootName, Namespace: *project.Spec.Namespace}
+
+		if err := gardenClient.Get(ctx, key, shoot); err != nil {
+			return nil, nil, nil, fmt.Errorf("invalid shoot %q: %w", key.Name, err)
+		}
+
+		return project, nil, shoot, nil
+	}
+
+	// list all shoots, filter by their name and possibly spec.seedName (if seed is set)
+	shootList := gardencorev1beta1.ShootList{}
+	if err := gardenClient.List(ctx, &shootList, &client.ListOptions{}); err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to list shoot clusters: %w", err)
+	}
+
+	// filter found shoots
+	matchingShoots := []*gardencorev1beta1.Shoot{}
+	for i, s := range shootList.Items {
+		if s.Name != shootName {
+			continue
+		}
+
+		// if filtering by seed, ignore shoot's whose seed name doesn't matchh
+		if seed != nil && (s.Status.SeedName == nil || *s.Status.SeedName != seed.Name) {
+			continue
+		}
+
+		matchingShoots = append(matchingShoots, &shootList.Items[i])
```

Comment:
why 
`matchingShoots = append(matchingShoots, &shootList.Items[i])`
and not 
`matchingShoots = append(matchingShoots, &s)`
?

Reply:
Because `s` is going to be reused in each loop interation. See https://play.golang.org/p/1YgZAv5JWNN -- using the loop variable's address is a common Go error that bit me many, many times.

Reply:
oops, bit me too! :D